src/visitor.rs

```diff
@@ -225,72 +199,99 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
-            let end_pos = source!(self, b.span).hi() - brace_compensation - remove_len;
-            let snippet = self.snippet(mk_sp(self.last_pos, end_pos));
-            snippet.contains("//") || snippet.contains("/*")
-        };
-        if unindent_comment {
+        let rest_span = self.next_span(b.span.hi());
+        if out_of_file_lines_range!(self, rest_span) {
+            self.push_str(self.snippet(rest_span));
             self.block_indent = self.block_indent.block_unindent(self.config);
+        } else {
+            // Ignore the closing brace.
+            let missing_span = self.next_span(b.span.hi() - brace_compensation);
+            self.close_block(missing_span, self.unindent_comment_on_closing_brace(b));
         }
-        self.format_missing_with_indent(
-            source!(self, b.span).hi() - brace_compensation - remove_len,
-        );
-        if unindent_comment {
-            self.block_indent = self.block_indent.block_indent(self.config);
-        }
-        self.close_block(unindent_comment, self.next_span(b.span.hi()));
         self.last_pos = source!(self, b.span).hi();
     }
 
-    // FIXME: this is a terrible hack to indent the comments between the last
-    // item in the block and the closing brace to the block's level.
-    // The closing brace itself, however, should be indented at a shallower
-    // level.
-    fn close_block(&mut self, unindent_comment: bool, span: Span) {
-        let skip_this_line = !self
-            .config
-            .file_lines()
-            .contains(&LineRange::from_span(self.source_map, span));
-        if skip_this_line {
-            self.push_str(self.snippet(span));
-        } else {
-            let total_len = self.buffer.len();
-            let chars_too_many = if unindent_comment {
-                0
-            } else if self.config.hard_tabs() {
-                1
-            } else {
-                self.config.tab_spaces()
-            };
+    fn close_block(&mut self, span: Span, unindent_comment: bool) {
+        let config = self.config;
 
-            // FIXME this is a temporaly fix
-            // should be remove truncate logic in close_block
-            // avoid not to truncate necessary chars
-            let truncate_start = total_len - chars_too_many;
-            let target_str = &self.buffer[truncate_start..total_len];
-            let truncate_length = target_str.len() - target_str.trim().len();
-
-            if let Some(last_char) = target_str.chars().last() {
-                self.buffer.truncate(total_len - truncate_length);
-                if last_char == '\n' {
-                    self.buffer.push_str("\n");
+        let mut last_hi = span.lo();
+        let mut unindented = false;
+        let mut prev_ends_with_newline = false;
+        let mut extra_newline = false;
+
+        let skip_normal = |s: &str| {
+            let trimmed = s.trim();
+            trimmed.is_empty() || trimmed.chars().all(|c| c == ';')
```

Comment:
Trailing semicolons are handled here.

Reply:
see above comment, to make sure this code is used, the source file should keep the trailing semicolons.